vector/src/main/java/im/vector/app/features/createdirect/CreateDirectRoomViewModel.kt

```diff
@@ -52,56 +56,151 @@ class CreateDirectRoomViewModel @AssistedInject constructor(@Assisted
+        // Tchap: - invites by msisdn are not supported yet
+        val selection = action.selections.singleOrNull() ?: return
+        setState { copy(isLoading = true) }
+        when (selection) {
+            // User already exists, so we can create or retrieve the DM with him
+            is PendingSelection.UserPendingSelection     -> handleCreateDirectMessageByUserId(selection.user.userId)
+            // User is unknown, so we have to invite him before creating the DM
+            is PendingSelection.ThreePidPendingSelection -> {
+                if (selection.threePid is ThreePid.Email) {
+                    handleIndividualInviteByEmail(selection.threePid.email)
+                } else {
+                    setState { copy(isLoading = false) }
+                    _viewEvents.post(CreateDirectRoomViewEvents.Failure(Throwable("Invite by Msisdn is not supported yet.")))
+                }
             }
-        } else {
-            // Create the DM
-            createRoomAndInviteSelectedUsers(action.selections)
         }
     }
 
-    private fun createRoomAndInviteSelectedUsers(selections: Set<PendingSelection>) {
-        setState { copy(createAndInviteState = Loading()) }
-
+    private fun handleCreateDirectMessageByUserId(userId: String) {
+        setState { copy(isLoading = true) }
         viewModelScope.launch(Dispatchers.IO) {
-            val adminE2EByDefault = rawService.getElementWellknown(session.sessionParams)
-                    ?.isE2EByDefault()
-                    ?: true
-
-            val roomParams = CreateRoomParams()
-                    .apply {
-                        selections.forEach {
-                            when (it) {
-                                is PendingSelection.UserPendingSelection     -> invitedUserIds.add(it.user.userId)
-                                is PendingSelection.ThreePidPendingSelection -> invite3pids.add(it.threePid)
-                            }.exhaustive
-                        }
-                        setDirectMessage()
-                        enableEncryptionIfInvitedUsersSupportIt = adminE2EByDefault
+            // Create or retrieve the DM and notify UI about the result
+            runCatching { directRoomHelper.ensureDMExists(userId) }.fold(
+                    {
+                        setState { copy(isLoading = false) }
+                        _viewEvents.post(CreateDirectRoomViewEvents.OpenDirectChat(it))
+                    },
+                    {
+                        setState { copy(isLoading = false) }
+                        _viewEvents.post(CreateDirectRoomViewEvents.Failure(it))
                     }
+            )
+        }
+    }
 
-            val result = runCatchingToAsync {
-                session.createRoom(roomParams)
+    private fun handleIndividualInviteByEmail(email: String) {
+        setState { copy(isLoading = true) }
```

Comment:
i think you don't need this line here, you already have an : " setState { copy(isLoading = true) }" line 73.

Reply:
This method is also called line 59, in this case the state will no be updated if I remove it

Reply:
Maybe you can move: "    setState { copy(isLoading = true) }" line 57, then you can delete it in: onSubmitInvitees, handleIndividualInviteByEmail and handleCreateDirectMessageByUserId

Reply:
I prefer to manage it separately.
Suppose we add an action which does not require a loading state, we'll be forced to deal with it. Also, at the beggining of onSubmitInvite, I don't update the state if the selection is bad, so I'll have to deal with it again